### PR TITLE
Fix getting started example

### PR DESCRIPTION
**Pull request by flxst**
The getting started example is updated as it does not reflect some recent changes in `modalities`. In particular, this PR aims to:

- [x] Give the readme file `examples/getting_started/README.md` a general update
- [x] Replace the previously employed tokenizer file (only `tokenizer.json`) by a tokenizer instance (e.g. the variant `PreTrainedHFTokenizer`)
- [x] Use dataset config files as input for the cli command `modalities data pack_encoded_data`
- [x] Update the training config file `examples/getting_started/example_config.yaml`
- [x] Get the cli command `modalities generate_text [..]` to work again (based on #102)

EDIT: Note that the above changes only make the getting started example work again. Suggestions for additional changes to improve user friendliness can be found in #117.

### CODE REVIEW

**Comment by flxst** [comment]
The getting started example has now been updated, adjusted to all the recent changes (CLI commands, config files etc.) and it's working again. However, the parameters in the training config file still need to be improved - they are currently not equivalent to those in the old training config file, and the text generated by the example model is very poor.

**Comment by flxst** [comment]
Parameters in the training and text generation config files have been improved. Text generated by the example model seems to be according to expectations.

File: examples/getting_started/README.md [hunk 1]
```diff
@@ -35,11 +38,38 @@ A single line of the Redpajama V2 JSONL file has the structure denoted below. Si
...
+  eod_token: <|endoftext|>
+
+tokenizer:
+  component_key: tokenizer
+  variant_key: pretrained_hf_tokenizer
+  config:
+    pretrained_model_name_or_path: tokenizer
+    padding: false
+    max_length: 1024
+````
```
Comment (le1nux) [inline examples/getting_started/README.md:62]:
```suggestion
```
```

Comment (le1nux) [reply examples/getting_started/README.md:62]: one too many ´

File: examples/getting_started/README.md [hunk 2]
```diff
@@ -35,11 +38,38 @@ A single line of the Redpajama V2 JSONL file has the structure denoted below. Si
 The two raw dataset splits for training and evaluation can be found in 
 `data/raw/redpajama_v2_samples_512_train.jsonl` and `data/raw/redpajama_v2_samples_512_test.jsonl`
 and need to be preprocessed into the [MemMap dataset format](https://github.com/Modalities/modalities/blob/main/src/modalities/dataloader/dataset.py). 
-Firstly, we create the dataset index via
 
+### Config File
+To do so, we employ the `example_dataset_config_train.yaml` and `example_dataset_config_test.yaml` configuration files, which contain the paths of the input and output files, the path of the tokenizer as well as some configurable parameters:
+```
```
Comment (le1nux) [inline examples/getting_started/README.md:44]:
```suggestion
```yaml
```

File: examples/getting_started/README.md [hunk 3]
```diff
@@ -35,11 +38,38 @@ A single line of the Redpajama V2 JSONL file has the structure denoted below. Si
...
+  num_cpus: ${node_env:num_cpus}
+  eod_token: <|endoftext|>
+
+tokenizer:
+  component_key: tokenizer
+  variant_key: pretrained_hf_tokenizer
+  config:
+    pretrained_model_name_or_path: tokenizer
+    padding: false
+    max_length: 1024
```
Comment (le1nux) [inline examples/getting_started/README.md:61]: for the example, maybe we want to go with  a context length of 512?

File: examples/getting_started/README.md [hunk 4]
```diff
@@ -162,16 +188,31 @@ The command can be broken down into the following parts:
...
+```
+# example_text_generation_config.yaml
+
+settings:
+  referencing_keys:
+    sample_key: input_ids
+    prediction_key: logits
+  model_path: ./checkpoints/<checkpoint_name>.bin
+  device: 0
+  context_length: 1024
```
Comment (le1nux) [inline examples/getting_started/README.md:204]: maybe 512?

File: examples/getting_started/example_config.yaml [hunk 5]
```diff
@@ -1,146 +1,247 @@
...
+    target_key: target_ids
+  training:
+    global_training_log_interval_in_steps: 48
+    global_checkpointing_interval_in_steps: 48
+    global_evaluation_interval_in_steps: 48
+    global_num_seen_steps: 0
+    do_apply_activation_checkpointing: true
+    gradient_acc_steps: 1
+    local_train_micro_batch_size: 8
+    sequence_length: 1024
```
Comment (le1nux) [inline examples/getting_started/example_config.yaml:14]: maybe 512?

File: examples/getting_started/example_config.yaml [hunk 6]
```diff
@@ -1,146 +1,247 @@
...
+  experiment_id: ${modalities_env:experiment_id}
+  referencing_keys:
+    sample_key: input_ids
+    target_key: target_ids
+  training:
+    global_training_log_interval_in_steps: 48
+    global_checkpointing_interval_in_steps: 48
+    global_evaluation_interval_in_steps: 48
+    global_num_seen_steps: 0
+    do_apply_activation_checkpointing: true
```
Comment (le1nux) [inline examples/getting_started/example_config.yaml:11]: for this small tutorial, I would suggest setting this to false

File: examples/getting_started/example_config.yaml [hunk 7]
```diff
@@ -1,146 +1,247 @@
...
-          config: 
-            raw_data_path: ./data/mem_map/redpajama_v2_samples_512_test.pbin
-            block_size: ${data.sequence_len}
-            sample_key: ${data.sample_key}
-        collate_fn:
-          type_hint: GPT2LLMCollator
-          config: 
-            sample_key: ${data.sample_key}
-            target_key: ${data.target_key}
+        batch_size: 4
```
Comment (le1nux) [inline examples/getting_started/example_config.yaml:94]: should we reference the batch size from the settings?

File: examples/getting_started/example_dataset_config_test.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  raw_samples_queue_size: 300
+  processed_samples_queue_size: 300
+
+tokenizer:
+  component_key: tokenizer
+  variant_key: pretrained_hf_tokenizer
+  config:
+    pretrained_model_name_or_path: tokenizer
+    padding: false
+    max_length: 1024
```
Comment (le1nux) [inline examples/getting_started/example_dataset_config_test.yaml:18]: maybe 512?

File: examples/getting_started/example_dataset_config_train.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  raw_samples_queue_size: 300
+  processed_samples_queue_size: 300
+
+tokenizer:
+  component_key: tokenizer
+  variant_key: pretrained_hf_tokenizer
+  config:
+    pretrained_model_name_or_path: tokenizer
+    padding: false
+    max_length: 1024
```
Comment (le1nux) [inline examples/getting_started/example_dataset_config_train.yaml:18]: maybe 512?

**Review by le1nux — APPROVED** [review]
Looks good to me! :) Just left a few minor remarks, worth thinking about but do not necessarily need to included.

File: examples/getting_started/README.md (hunk 3 above)
Comment (flxst) [reply examples/getting_started/README.md]: Yes, I initially chose 1024 here to differentiate this from the number of documents in the example data, which is also 512. But I changed the context length back to 512 now in e391cfc2646dcb52e9943449cc42f665f3b3b5fc.

File: examples/getting_started/README.md (hunk 4 above)
Comment (flxst) [reply examples/getting_started/README.md]: Fixed in e391cfc2646dcb52e9943449cc42f665f3b3b5fc.

File: examples/getting_started/example_config.yaml (hunk 5 above)
Comment (flxst) [reply examples/getting_started/example_config.yaml]: Fixed in e391cfc2646dcb52e9943449cc42f665f3b3b5fc.

File: examples/getting_started/example_dataset_config_test.yaml (hunk 8 above)
Comment (flxst) [reply examples/getting_started/example_dataset_config_test.yaml]: Fixed in e391cfc2646dcb52e9943449cc42f665f3b3b5fc.

File: examples/getting_started/example_dataset_config_train.yaml (hunk 9 above)
Comment (flxst) [reply examples/getting_started/example_dataset_config_train.yaml]: Fixed in e391cfc2646dcb52e9943449cc42f665f3b3b5fc.

File: examples/getting_started/example_config.yaml (hunk 6 above)
Comment (flxst) [reply examples/getting_started/example_config.yaml]: Fixed in bd8ca46c3f9691dfa710ed5852ceb1334359a33e.

File: examples/getting_started/example_config.yaml (hunk 7 above)
Comment (flxst) [reply examples/getting_started/example_config.yaml]: Yes, I changed it to `${settings.training.local_train_micro_batch_size}` in 6c86006e5c1f6efa6ea8739ffe2931919d21441a. 

Note, however, that in `config_files/training/config_lorem_ipsum.yaml`, the `batch_size` is hardcoded as well (I did not change this as it is not related to the getting started example). Also, if we decide to always use the same `batch_size` for training, validation and testing, we might want to rename `local_train_micro_batch_size` to `local_micro_batch_size`. I will leave this "for future work" :)
